### PR TITLE
[ready] Layer Normalization

### PR DESCRIPTION
**Pull request by ssnl**
Commits:
1. Renames `ATen/Check.h` to `ATen/TensorUtils.h` with an additional method `at::maybe_data_ptr` added. #4851 
2. THNN BN code so that `running_mean` and `running_var` can be optional when `training=True`. #4509 
3. ATen changes and cuDNN changes. Still #4509 
4. Python nn.* changes, including changing `InstanceNorm*d`'s `use_running_stats` from #4444 to the new option `track_running_stats` on BN. Improves IN and BN docs.
5. Adds test for the new option for IN and BN. Improves other IN tests.
6. Adds Layer Normalization #1959  .
7. Fixes LRN doc.
8. Functional interface for IN and LN. 
9. Tests for LN. 
10. Fix BN double backward returning undefined tensor when it shouldn't.
11. Fix Jit tests that use wrong dim inputs for BN
12. Add/Improve BN, IN and LN GPU tests with half.
13. Update IN BN docs to be consistent with conv notation; Fix onnx failures.

### CODE REVIEW

File: torch/nn/modules/layernorm.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class LayerNorm(_BatchNorm):
+    r"""Applies Layer Normalization over a mini-batch of inputs as described in
+    the paper `Layer Normalization`_ .
+
+    .. math::
+        y = \frac{x - mean[x]}{ \sqrt{Var[x]} + \epsilon} * gamma + beta
+
+    The mean and standard-deviation are calculated separately for all dimensions
```
Comment (jekbradbury) [inline torch/nn/modules/layernorm.py]: This is a little unclear. What layernorm should do, as far as I know, is compute means and stdevs over a single dimension (representing the channel, typically dimension -1) or occasionally in the case of image data multiple dimensions (i.e. one or more spatial dims + the channel dim). But at least for the NLP use case it shouldn't reduce over all non-batch dimensions, only one of them, and it should never compute statistics over the batch dimension.

Comment (SsnL) [reply torch/nn/modules/layernorm.py]: From the way the original paper describes the concept and formulates LN, LN normalizes all inputs to a layer. So it's across dimensions. tf.contrib implementation also agrees on this https://www.tensorflow.org/api_docs/python/tf/contrib/layers/layer_norm. That said, it is not quite clear what LN does when input has channels/planes (e.g. images) since it makes more sense to have it behave like IN. @Kaixhin and I had a discussion on this, and we hoped to get a concrete answer from the author. Unfortunately, we haven't heard back.

Comment (SsnL) [reply torch/nn/modules/layernorm.py]: And yeah, it doesn't compute statistics across the batch dimension. The implementation doesn't do so. I need to fix the doc :)

Comment (jekbradbury) [reply torch/nn/modules/layernorm.py]: The TF implementation has a parameterizable number of dimensions to normalize over; it always normalizes over the last `n` dimensions but `n` is configurable. That seems like an OK API, since the channel dimension is virtually always last.
The original paper normalizes over "all the hidden units in a layer," which is a little ambiguous, but clarifies that in recurrent networks the normalization is performed "at the current timestep," which would translate to normalizing only over the channel dimension if the input tensor also had a time dimension; that's the strategy adopted in e.g. [the Transformer network](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/common_layers.py#L517).

Comment (SsnL) [reply torch/nn/modules/layernorm.py]: Your argument on RNN is valid. The problem with RNN is that it has two aggregating dimensions (batch and time). It's my bad for forgetting LN's popularity in NLP and focusing only on use cases similar to BN and IN. I believe an API like the tf.contrib one is reasonable in this case. 

However, what is the difference between IN and LN then? From the user's perspective, what LN can do is a superset of what IN can do.

Comment (Kaixhin) [reply torch/nn/modules/layernorm.py]: In TF the order for images is `NHWC` but for us it's `NCHW` (same type of ordering for 1D and 3D signals), so we can't copy their API directly. For RNNs which take an entire sequence, we have `TNC` by default, or `NTC` if `batch_first = True`. I may have got this wrong, but LN/IN (which I'm considering the same atm, with different use-cases explained in the docs) should choose the first dim (from the left) to normalise, and then collapse everything to the right into this dim?

**Comment by Kaixhin** [comment]
Sorry but could you also fix the LRN docs with this PR? Just need to remove the extra colon [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/normalization.py#L25).

File: torch/nn/modules/layernorm.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/modules/layernorm.py]: @Kaixhin Yes, that is what I'm thinking about as well. (It is essentially the tf API btw.)

Comment (Kaixhin) [reply torch/nn/modules/layernorm.py]: Great so we have something like `begin_norm_axis` but can get rid of `begin_params_axis` because of our ordering?

Comment (SsnL) [reply torch/nn/modules/layernorm.py]: Yes. I plan to only have `begin_norm_axis` (maybe with a different name).

**Comment by SsnL** [comment]
This is ready for review. The windows build failure are unrelated known `torch.Size` issues on master. The onnx-fb-universe failure is also unrelated.

File: torch/nn/modules/instancenorm.py [hunk 2]
```diff
@@ -3,71 +3,62 @@
...
-        """
-        self._use_running_stats = mode
+        return F.instance_norm(
+            input, self.running_mean, self.running_var, self.weight, self.bias,
+            self.training or not self.track_running_stats, self.momentum, self.eps)
 
 
 class InstanceNorm1d(_InstanceNorm):
-    r"""Applies Instance Normalization over a 3d input that is seen as a mini-batch.
+    r"""Applies Batch Normalization over a 3d input (a mini-batch of 2d inputs)
```
Comment (Kaixhin) [inline torch/nn/modules/instancenorm.py]: Typo here as well as at the 2D and 3D versions.

**Comment by Kaixhin** [comment]
So we've gone for LayerNorm normalising over every element in each sample, and InstanceNorm{1,2,3}d normalising only over channels?

**Comment by SsnL** [comment]
Well, LN normalizes the last certain number dimensions. And IN normalizes
the dimensions after channel. So LN can do what IN does in terms of
normalization. However, they apply affine transforms differently.

**Comment by Kaixhin** [comment]
Pinging @DmitryUlyanov in case he wants to check the refactoring of InstanceNorm.

**Comment by ngimel** [comment]
Can you please add a smoke test for half, like Sam did for batch norm? https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3317-L3324
Normal tests don't test half. 
Also, not to delay this PR, but it would make sense to move pytorch batchnorm (that is used here) to use Welford for variance computation.

**Comment by SsnL** [comment]
@nigmel I will add a smoke test. Thanks for the reminder!

About the Welford variance algorithm in THNN BN path, I can make a separate
PR for that since it is basically orthogonal with this one.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -1142,54 +1142,124 @@ def embedding_bag(embedding_matrix, indices, offsets=None,
...
 
-def instance_norm(input, weight, bias, saved_running_mean, saved_running_var,
-                  training=False, momentum=0.1, eps=1e-5, affine=False):
-    """Applies instance normalization over an input. The implementation is
-    based on batch_norm, in which we do reshape, batchnorm, and reshape again.
+def batch_norm(input, running_mean, running_var, weight=None, bias=None,
+               training=False, momentum=0.1, eps=1e-5):
+    """Applies Batch Normalization for each channel across a batch of data.
+
+    See :class:`~torch.nn.BatchNorm2d` (or the 1d and 3d variants) for details.
```
Comment (Kaixhin) [inline torch/nn/functional.py]: For consistency with `nn.InstanceNorm` maybe we want the following?

> See :class:`~torch.nn.BatchNorm1d`, :class:`~torch.nn.BatchNorm2d`, :class:`~torch.nn.BatchNorm3d` for details.

File: torch/nn/modules/batchnorm.py [hunk 4]
```diff
@@ -8,43 +8,56 @@
...
         return ('{name}({num_features}, eps={eps}, momentum={momentum},'
-                ' affine={affine})'
+                ' affine={affine}, track_running_stats={track_running_stats})'
                 .format(name=self.__class__.__name__, **self.__dict__))
 
 
 class BatchNorm1d(_BatchNorm):
-    r"""Applies Batch Normalization over a 2d or 3d input that is seen as a
-    mini-batch.
+    r"""Applies Batch Normalization over a 3d input (a mini-batch of 2d inputs)
```
Comment (Kaixhin) [inline torch/nn/modules/batchnorm.py]: Is this not still 2D and 3D? This was overloaded to do the original batch norm with N x C (e.g. fully-connected layers) and the 1D signal version N x C x L (e.g. 1D conv layers).

File: torch/nn/modules/batchnorm.py [hunk 5]
```diff
@@ -54,23 +67,39 @@ class BatchNorm1d(_BatchNorm):
...
+        where :math:`\hat{x}` is the estimated statistic and :math:`x_t` is the
+        new observed value.
 
     Because the BatchNorm is done over the `C` dimension, computing statistics
     on `(N, L)` slices, it's common terminology to call this Temporal BatchNorm
 
     Args:
         num_features: num_features from an expected input of size
-            `batch_size x num_features [x width]`
+            `[batch_size x num_features (x width)]`
```
Comment (Kaixhin) [inline torch/nn/modules/batchnorm.py]: For consistency with convolutions this should be `length` for 1D signals (L for 1D, H x W for 2D, D x H x W for 3D).

File: torch/nn/modules/instancenorm.py [hunk 6]
```diff
@@ -3,61 +3,62 @@
...
+        classes and the conventional notion of momentum. Mathematically, the
+        update rule for running statistics here is
+        :math:`\hat{x}_\text{new} = (1 - \text{momentum}) \times \hat{x}_\text{new} + \text{momemtum} \times x_t`,
+        where :math:`\hat{x}` is the estimated statistic and :math:`x_t` is the
+        new observed value.
 
     Args:
-        num_features: num_features from an expected input of size `batch_size x num_features x width`
+        num_features: num_features from an expected input of size
+            `[batch_size x num_features x width]`
```
Comment (Kaixhin) [inline torch/nn/modules/instancenorm.py]: Should be `length` for consistency.

**Comment by Kaixhin** [comment]
Left a few notes on docs but the rest LGTM - thanks a lot @SsnL!

**Comment by SsnL** [comment]
Yay! Finally fixed the onnx test.

Now that the tests should pass. I can confidently ask for reviews.

@jekbradbury Does the new proposed interface look good to you? Specifically, it is `nn.LayerNorm(normalized_shape)`, where `normalized_shape` is the shape of the last few dimensions to be normalized over. The shape info is needed to initialize the affine transform parameters which Layer Norm is usually used together with.

**Comment by jekbradbury** [comment]
Yes, this looks great!

File: torch/nn/modules/normalization.py [hunk 7]
```diff
@@ -69,6 +72,142 @@ def __repr__(self):
...
+        This :attr:`momentum` argument is different from one used in optimizer
+        classes and the conventional notion of momentum. Mathematically, the
+        update rule for running statistics here is
+        :math:`\hat{x}_\text{new} = (1 - \text{momentum}) \times \hat{x}_\text{new} + \text{momemtum} \times x_t`,
+        where :math:`\hat{x}` is the estimated statistic and :math:`x_t` is the
+        new observed value.
+
+    Args:
+        normalized_shape (list or torch.Size): input shape from an expected input of size
+            `[* x normalized_shape[0] x normalized_shape[1] x ... x normalized_shape[-1]]`
```
Comment (zou3519) [inline torch/nn/modules/normalization.py]: Nit: you can add a comment saying that * represents any number of dimensions (though we do have it elsewhere in our docs)

File: aten/src/ATen/TensorUtils.h [hunk 8]
```diff
@@ -4,14 +4,14 @@
...
-// This file contains utility functions for checking that arguments
-// make sense.  This is particularly useful for native functions,
-// which do NO argument checking by default.
-//
-// It's NOT in Utils.h, because this file has a dep on Tensor.h
+// These functions are NOT in Utils.h, because this file has a dep on Tensor.h
 
 namespace at {
 
+// The follwing are utility functions for checking that arguments
```
Comment (zou3519) [inline aten/src/ATen/TensorUtils.h]: nit: following

File: aten/src/THCUNN/BatchNormalization.cu [hunk 9]
```diff
@@ -141,8 +141,8 @@ template <typename Dtype, typename Acctype, typename DeviceTensor1, typename Dev
 __global__ void BatchNormalizationUpdateOutputInference_kernel(
     const DeviceTensor3 input,
     DeviceTensor3 output,
-    DeviceTensor1 runningMean,
-    DeviceTensor1 runningVar,
+    const DeviceTensor1 runningMean,
+    const DeviceTensor1 runningVar,
```
Comment (zou3519) [inline aten/src/THCUNN/BatchNormalization.cu]: Why are you marking these as const?

Comment (SsnL) [reply aten/src/THCUNN/BatchNormalization.cu]: good point! will fix.

Comment (SsnL) [reply aten/src/THCUNN/BatchNormalization.cu]: Oh that's because that this is the inference forward path, which means that the running stats will not be updated.

File: test/test_nn.py [hunk 10]
```diff
@@ -3539,6 +3591,30 @@ def _test_batchnorm_eval(self, test_type=torch.FloatTensor):
         self.assertEqual(res1, res2)
         self.assertEqual(grad1, grad2)
 
+        # track_running_stats=False
+        module = nn.BatchNorm1d(3, track_running_stats=False).type(test_type)
```
Comment (zou3519) [inline test/test_nn.py]: Do you test track_running_stats=False for BatchNorm{2d,3d}?

File: torch/nn/functional.py [hunk 11]
```diff
@@ -1171,66 +1171,124 @@ def embedding_bag(embedding_matrix, indices, offsets=None,
     return ret
 
 
-def instance_norm(input, weight, bias, saved_running_mean, saved_running_var,
-                  training=False, momentum=0.1, eps=1e-5, affine=False):
-    """Applies instance normalization over an input. The implementation is
-    based on batch_norm, in which we do reshape, batchnorm, and reshape again.
+def batch_norm(input, running_mean, running_var, weight=None, bias=None,
+               training=False, momentum=0.1, eps=1e-5):
```
Comment (zou3519) [inline torch/nn/functional.py]: Is there a test to check if running_mean and running_var are updated?

File: torch/nn/modules/batchnorm.py [hunk 12]
```diff
@@ -89,7 +89,7 @@ class BatchNorm1d(_BatchNorm):
 
     Args:
         num_features: num_features from an expected input of size
-            `[batch_size x num_features (x width)]`
+            :math:`(N, C, L)` or :math:`(N, L)`
```
Comment (zou3519) [inline torch/nn/modules/batchnorm.py]: reminder to clarify this

File: torch/nn/modules/instancenorm.py [hunk 13]
```diff
@@ -50,7 +50,7 @@ class InstanceNorm1d(_InstanceNorm):
 
     Args:
         num_features: num_features from an expected input of size
-            `[batch_size x num_features x width]`
+            :math:`(N, C, L)` or :math:`(N, L)`
```
Comment (zou3519) [inline torch/nn/modules/instancenorm.py]: ditto

File: torch/nn/functional.py (hunk 11 above)
Comment (zou3519) [reply torch/nn/functional.py]: resolved offline

File: test/test_nn.py (hunk 10 above)
Comment (SsnL) [reply test/test_nn.py]: After discussion offline, we notice that this method originally only checks for 1d case. Considering that gradchecks for this case is done below, and that all three use the same underlying implementation, this should be fine.

**Review by zou3519 — APPROVED** [review]
LGTM!

**Comment by ruotianluo** [comment]
The docs of Layernorm and LRN has these:
are these expected?

http://pytorch.org/docs/master/nn.html?highlight=layernorm#torch.nn.LayerNorm.forward
http://pytorch.org/docs/master/nn.html?highlight=layernorm#torch.nn.LocalResponseNorm.forward

**Comment by SsnL** [comment]
@ruotianluo definitely not. This phenomenon is not unique to these two class through. I'll figure out what's wrong.

**Comment by meder411** [comment]
Is there a particular reason that the `normalized_shape` input must be a tensor? For fully connected layers or other 1D inputs, it seems that the constructor ought to be able to accept an integer. As of now, I am using 

`nn.LayerNorm(torch.LongTensor([256])`

From the paper, there doesn't seem to be any reason you wouldn't use Layer Normalization in this way. It's a minor change, but it may be a more seemless interface.

**Comment by SsnL** [comment]
@meder411 It doesn't need to be a tensor. It can be a list, tuple, `torch.Size`, etc., basically anything that can be accepted by `torch.Size`'s constructor. In fact, the doc never suggested passing in a tensor, so I'm not sure where you get that idea.

The idea is to be able to normalize over multiple dimensions. Hence the argument is called `normalized_shape`. The doc explains it pretty clearly I think.

**Comment by Kaixhin** [comment]
@meder411 While the docs do look correct to me for describing `normalized_shape`, having it accept an integer for the most common use-case by far does seem worthwhile. If you're willing to submit a PR for this that'd be great, otherwise raise an issue and someone will address it.

File: torch/nn/modules/normalization.py [hunk 14]
```diff
@@ -69,6 +71,109 @@ def __repr__(self):
...
+            self.register_parameter('running_mean', None)
+            self.register_parameter('running_var', None)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.track_running_stats:
+            self.running_mean.zero_()
+            self.running_var.fill_(1)
+        if self.elementwise_affine:
+            self.weight.data.uniform_()
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/nn/modules/normalization.py]: I'm following how BN and IN initializes weights here.

[the remark this answers is not in the document]
Comment (avati) [reply torch/nn/modules/normalization.py]: Going by the paper, this should be initialized to 1 and not randomly initialized. Think that should be the case for BN as well.

[the remark this answers is not in the document]
Comment (SsnL) [reply torch/nn/modules/normalization.py]: Fair point. I'll submit a PR for these.

**Comment by JustinLin610** [comment]
How can I use it with nn.LSTM or nn.GRU? it will be much more helpful if LN can be used with the two classes

**Comment by SsnL** [comment]
@JustinLin610 Using the `*Cell` classes is the only way.

**Comment by jinserk** [comment]
I'd like to apply LayerNorm to multi-layer LSTM but have no idea how to use LSTMCell classes. @SsnL, could you let me know any simple example for this? LayerNorm is quite new to PyTorch so it is difficult to find a good example for this..

**Comment by MBAnslow** [comment]
@ssnl I agree with jinserk. I've looked far and wide for a feature full example of LSTMs and GRUs that support layer normalisation including bidirectionality and multiple layers and I haven't really found anything. There are some incomplete LSTM versions around but they don't seem to be feature complete or optimised very well.

**Comment by zou3519** [comment]
@MBAnslow, @jinserk  see https://github.com/pytorch/benchmark/blob/master/rnns/fastrnns/custom_lstms.py for some examples
